astropy/io/ascii/ecsv.py

```diff
@@ -203,10 +229,79 @@ def __call__(self, cols, meta):
+        for col in cols:
+            try:
+                # 1-d or N-d object columns are serialized as JSON.
+                if col.subtype == 'object':
+                    _check_dtype_is_str(col)
+                    col_vals = [json.loads(val) for val in col.str_vals]
+                    col.data = np.empty([len(col_vals)] + col.shape, dtype=object)
+                    col.data[...] = col_vals
+
+                # Variable length arrays with shape (n, m, ..., *) for fixed
+                # n, m, .. and variable in last axis. Masked values here are
+                # not currently supported.
+                elif col.shape and col.shape[-1] is None:
+                    _check_dtype_is_str(col)
+                    # Remake as a 1-d object column of numpy ndarrays using the
+                    # datatype specified in the ECSV file.
+                    col_vals = [np.array(json.loads(val), dtype=col.subtype)
+                                for val in col.str_vals]
+                    col.shape = ()
+                    col.dtype = np.dtype(object)
+                    # np.array(col_vals_arr, dtype=object) fails ?? so this workaround:
+                    col.data = np.empty(len(col_vals), dtype=object)
+                    col.data[:] = col_vals
+
+                # Multidim columns with consistent shape (n, m, ...). These
+                # might be masked.
+                elif col.shape:
+                    _check_dtype_is_str(col)
+                    col_vals = [json.loads(val) for val in col.str_vals]
+                    # Make a numpy object array of col_vals to look for None
+                    # (masked values)
+                    data = np.array(col_vals, dtype=object)
+                    mask = (data == None)  # noqa: E711
+                    if not np.any(mask):
+                        # No None's, just convert to required dtype
+                        col.data = data.astype(col.subtype)
+                    else:
+                        # Replace all the None with an appropriate fill value
+                        kind = np.dtype(col.subtype).kind
+                        data[mask] = {'U': '', 'S': b''}.get(kind, 0)
+                        # Finally make a MaskedArray with the filled data + mask
+                        col.data = np.ma.array(data.astype(col.subtype), mask=mask)
+
+                # Regular scalar value column
+                else:
+                    if col.subtype:
+                        warnings.warn(f'unexpected subtype {col.subtype!r} set for column '
+                                      f'{col.name!r}, using dtype={col.dtype!r} instead.',
+                                      category=AstropyUserWarning)
+                    converter_func, _ = convert_numpy(col.dtype)
+                    col.data = converter_func(col.str_vals)
+
+                if col.data.shape[1:] != tuple(col.shape):
+                    raise ValueError('shape mismatch between value and column specifier')
```

Comment:
Not sure we can get away with just a warning here...

Regardless, maybe not have this inside the `try/except`?